gcp_variant_transforms/transforms/infer_headers.py

```diff
@@ -310,12 +312,96 @@ def expand(self, pcoll):
             # be used. Should this changes, we should modify the default value.
             | 'MergeHeaders' >> merge_headers.MergeHeaders(
                 split_alternate_allele_info_fields=True,
                 allow_incompatible_records=self._allow_incompatible_records))
+
+
+class _InferAnnotationTypes(_InferHeaderFields):
+  """Calculates types of annotation fields from `Variant` records.
+
+  All annotation headers in each annotation field are converted to Info header
+  lines where the new ID corresponds to the given annotation field and header,
+  and the new TYPE corresponds to inferred type of the original header. Since
+  each variant potentially contains multiple values for each annotation header,
+  a small 'merge' of value types is performed before VcfHeader creation for each
+  variant.
+  """
+
+  def process(self,
+              variant,  # type: vcfio.Variant
+              defined_headers,  # type: vcf_header_io.VcfHeader
+              annotation_fields  # type: List[str]
+             ):
+    # type: (...) -> Iterable[vcf_header_io.VcfHeader]
+    """
+    Args:
+      defined_headers: header fields defined in header section of VCF files.
+      annotation_fields: list of info fields treated as annotation fields
+        (e.g. ['CSQ_VT'])
+
+    Yields:
+      inferred_header_infos: A VcfHeader with info fields documenting newly
+        inferred annotation types
+    """
+    infos = {}
+    resolver = vcf_field_conflict_resolver.FieldConflictResolver(
+        resolve_always=True)
+    for field in annotation_fields:
+      if field not in variant.info:
+        continue
+      annotation_headers = annotation_parser.extract_annotation_names(
+          defined_headers.infos[field][
+              vcf_header_io.VcfParserHeaderKeyConstants.DESC], with_alt=True)
+      annotation_values = [annotation_parser.extract_annotation_list_with_alt(
+          annotation) for annotation in variant.info[field]]
+      annotation_values = zip(*annotation_values)
+      for header, values in zip(annotation_headers, annotation_values):
+        variant_merged_type = vcf_header_io.VcfHeaderFieldTypeConstants.INTEGER
+        for v in values:
+          if v != '':
+            variant_merged_type = resolver.resolve_attribute_conflict(
+                vcf_header_io.VcfParserHeaderKeyConstants.TYPE,
+                variant_merged_type,
+                self._get_field_type(v))
+        key_id = vcf_header_io.BASE_TYPE_KEY.format(field, header)
+        infos[key_id] = Info(key_id,
+                             1,
+                             variant_merged_type,
+                             '',  # NO_DESCRIPTION
+                             '',  # UNKNOWN_SOURCE
+                             '')  # UNKNOWN_VERSION
+    inferred_header_infos = vcf_header_io.VcfHeader(infos=infos)
+    yield inferred_header_infos
+
+
+class InferAnnotationTypes(beam.PTransform):
```

Comment:
Please see my performance comment in `vcf_to_bq.py` and verify what I am saying is actually correct. To fix that multiple pass issue (i.e., going twice through all variants to infer headers), I think you can implement the logic you have in this `PTransform` inside `InferHeaderFields` instead. For example, one way that I can think of fixing this is by changing [`_infer_info_fields` method](https://github.com/googlegenomics/gcp-variant-transforms/blob/10aa5f8632f7b750fc2cf5ebcd434fb04566a29f/gcp_variant_transforms/transforms/infer_headers.py#L182) and add the logic for annotation header field generation in there. Then you can avoid the second pass and do everything in one `InferHeaderFields` pass.